File: account/src/main/java/serp/project/account/ui/controller/RoleController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import serp.project.account.core.domain.dto.request.AddPermissionToRoleDto;
import serp.project.account.core.domain.dto.request.CreateRoleDto;
import serp.project.account.core.domain.dto.request.UpdateRoleDto;
import serp.project.account.core.usecase.RoleUseCase;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/roles")
public class RoleController {
    private final RoleUseCase roleUseCase;

    @PostMapping
    public ResponseEntity<?> createRole(@Valid @RequestBody CreateRoleDto createRoleDto) {
        var response = roleUseCase.createRole(createRoleDto);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/{roleId}/permissions")
    public ResponseEntity<?> addPermissionsToRole(
            @PathVariable Long roleId,
            @Valid @RequestBody AddPermissionToRoleDto request) {
        var response = roleUseCase.addPermissionsToRole(roleId, request.getPermissionIds());
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getAllRoles() {
        var response = roleUseCase.getAllRoles();
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/{roleId}")
    public ResponseEntity<?> updateRole(
            @PathVariable Long roleId,
            @Valid @RequestBody UpdateRoleDto request) {
        var response = roleUseCase.updateRole(roleId, request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

}

File: api_gateway/src/ui/controller/account/role_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type RoleController struct {
	roleService service.IRoleService
}

func (r *RoleController) CreateRole(c *gin.Context) {
	var req request.CreateRoleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := r.roleService.CreateRole(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (r *RoleController) GetAllRoles(c *gin.Context) {
	res, err := r.roleService.GetAllRoles(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (r *RoleController) AddPermissionsToRole(c *gin.Context) {
	roleId, ok := utils.ValidateAndParseID(c, "roleId")
	if !ok {
		return
	}

	var req request.AddPermissionToRoleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := r.roleService.AddPermissionsToRole(c.Request.Context(), roleId, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (r *RoleController) UpdateRole(c *gin.Context) {
	roleId, ok := utils.ValidateAndParseID(c, "roleId")
	if !ok {
		return
	}

	var req request.UpdateRoleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := r.roleService.UpdateRole(c.Request.Context(), roleId, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewRoleController(roleService service.IRoleService) *RoleController {
	return &RoleController{
		roleService: roleService,
	}
}
